packages/frontend/i18n/src/index.ts

```diff
@@ -35,88 +33,87 @@ declare module 'react-i18next' {
   }
 }
 
-const STORAGE_KEY = 'i18n_lng';
-
-export { I18nextProvider, LOCALES, Trans };
-
-const resources = LOCALES.reduce<Resource>((acc, { tag, res }) => {
-  return Object.assign(acc, { [tag]: { translation: res } });
-}, {});
+const logger = new DebugLogger('i18n');
 
-const fallbackLng = 'en';
-const standardizeLocale = (language: string) => {
-  if (language === 'zh-CN' || language === 'zh' || language === 'zh-Hans') {
-    language = 'zh-Hans';
-  } else if (language.slice(0, 2).toLowerCase() === 'zh') {
-    language = 'zh-Hant';
-  }
-  if (LOCALES.some(locale => locale.tag === language)) return language;
-  if (
-    LOCALES.some(locale => locale.tag === language.slice(0, 2).toLowerCase())
-  ) {
-    return language.slice(0, 2).toLowerCase();
-  }
+const defaultLng: Language = 'en';
 
-  return fallbackLng;
-};
+let _instance: i18n | null = null;
+export const getOrCreateI18n = (): i18n => {
+  if (!_instance) {
+    _instance = i18next.createInstance();
+    _instance
+      .use(initReactI18next)
+      .use({
+        type: 'backend',
+        init: () => {},
+        read: (lng: Language, _ns: string, callback) => {
+          const resource = SUPPORTED_LANGUAGES[lng].resource;
+          if (typeof resource === 'function') {
+            resource()
+              .then(data => {
+                logger.info(`Loaded i18n ${lng} resource`);
+                callback(null, data);
+              })
+              .catch(err => {
+                logger.error(`Failed to load i18n ${lng} resource`, err);
```

Comment:
fallback to `defaultLng` when load error?

Reply:
fallback setting at L67, 'en' is always loaded

Reply:
i think `fallbackLng` at `L67` is for individual missing trans keys for that locale.

Reply:
the resolver will always traverse all fallbacks to find a key

```
langs = ['zh-Hans', 'en']

foreach lang in langs
  if lang.resolve(key) 
    break
```